### PR TITLE
[AVRO-2677] Added decimal encoding and decoding logic

### PR DESCRIPTION
**Pull request by rpbaltazar**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by Fokko** [comment]
@rpbaltazar Can you rebase onto master?

**Comment by rpbaltazar** [comment]
Everything seems to be synced on my side. Not sure what you're referring to.

**Comment by Fokko** [comment]
@rpbaltazar My mistake, we had some issues with a broken master. I've restarted the failing CI jobs.

File: lang/ruby/lib/avro/logical_types.rb [hunk 1]
```diff
@@ -61,6 +61,16 @@ def self.decode(int)
       end
     end
 
+    module Decimal
```
Comment (tjwp) [inline lang/ruby/lib/avro/logical_types.rb]: For consistency here, I think this module should be called `BytesDecimal`.

**Comment by ziggythehamster** [comment]
> I believe that the encoded representation of the decimal should be the unscaled value, not a string representation.

Even worse, you have to produce a special sequence of bytes (which in Ruby is a string, but don't let that fool you into thinking `"1.234"` is correct). It must conform to this specification:

> A decimal logical type annotates Avro bytes or fixed types. The byte array must contain the two's-complement representation of the unscaled integer value in big-endian byte order. The scale is fixed, and is specified using an attribute.

https://avro.apache.org/docs/current/spec.html#Decimal

(copied from the other ticket)

**Comment by ziggythehamster** [comment]
Here's the Java decoder: https://github.com/apache/avro/blob/93ec41bacf678a875d3696dec72a84901482602a/lang/java/avro/src/main/java/org/apache/avro/Conversions.java#L82-L88

and the Java encoder: https://github.com/apache/avro/blob/93ec41bacf678a875d3696dec72a84901482602a/lang/java/avro/src/main/java/org/apache/avro/Conversions.java#L90-L95

and since I'm not a Java programmer, I looked at the docs for Java BigDecimal and see this statement:

> The value of the number represented by the BigDecimal is therefore (unscaledValue × 10-scale).

Excellent, because Ruby's [BigDecimal#split](https://ruby-doc.org/stdlib-2.7.1/libdoc/bigdecimal/rdoc/BigDecimal.html#method-i-split) method will give you the significant digits and exponent which I think is the same.

Of course, you now need to figure out what Java's BigInteger `toByteArray` is doing exactly, and [here's the docs for that](https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html#toByteArray--). Almost a decade ago, someone on Stack Overflow worked on a [version of this](https://stackoverflow.com/questions/5284369/ruby-return-byte-array-containing-twos-complement-representation-of-bignum-fi)... and it's been longer than that since I took any CS classes, so hopefully someone with a bit more freshness there can take this information and run with it :).

**Comment by rpbaltazar** [comment]
Since there is a PR (#829) open with the similar changes and i don't have the time to work on this, i'll close my pr. If i find some time later on i can try to help but don't want to commit atm.

**Pull request closed without merging**